src/jupyterlabcontroller/services/lab.py

```diff
@@ -287,3 +588,591 @@ async def _gather_secret_data(
+            state = await self._builder.recreate_lab_state(username, objects)
+            lab = self._labs.get(username)
+            if lab and lab.monitor.in_progress:
+                continue
+            if state:
+                observed[username] = state
+            elif not lab:
+                msg = "Deleting incomplete namespace"
+                self._logger.warning(msg, user=username, namespace=namespace)
+                await self._storage.delete_namespace(namespace)
+
+        # If the set of users we expected to see changed during
+        # reconciliation, that means someone added a new user while we were
+        # reconciling. Play it safe and skip this background update; we'll
+        # catch any inconsistencies the next time around.
+        #
+        # From this point forward, make sure not to do any asyncio operations
+        # until we've finished reconciling state, since if we yield control
+        # our state may change out from under us.
+        if set(self._labs.keys()) != known_users:
+            msg = "Known users changed during reconciliation, skipping"
+            self._logger.info(msg)
+            return
+
+        # First pass: check all users already recorded in internal state
+        # against Kubernetes and correct them (or remove them) if needed.
+        for username, lab in self._labs.items():
+            if lab.monitor.in_progress:
+                continue
+            if not lab.state:
+                continue
+            if lab.state.status == LabStatus.FAILED:
+                continue
+            if username not in observed:
+                msg = f"Expected user {username} not found in Kubernetes"
+                self._logger.warning(msg)
+                lab.state.status = LabStatus.FAILED
+            else:
+                observed_state = observed[username]
+                if observed_state.status != lab.state.status:
+                    self._logger.warning(
+                        f"Expected status for {username} is"
+                        f" {lab.state.status}, but observed status is"
+                        f" {observed_state.status}"
+                    )
+                    lab.state.status = observed_state.status
+
+                # If we discovered the pod was actually in pending state, kick
+                # off a monitoring job to wait for it to become ready and
+                # handle timeouts if it never does.
+                if observed_state.status == LabStatus.PENDING:
+                    to_monitor.append(username)
+
+        # Second pass: take observed state and create any missing internal
+        # state. This is the normal case after a restart of the lab
+        # controller.
+        for username in set(observed.keys()) - known_users:
+            if username not in self._labs:
+                msg = f"Creating record for user {username} from Kubernetes"
+                self._logger.info(msg)
+                self._labs[username] = _State(
```

Comment:
Does all of this get all gross again when we push state into an external redis, which we would presumably want to access asynchonously?

Reply:
Yes. We'll have to completely redesign the reconciliation pass when we use Redis for state.